qadence/backends/pyqtorch/convert_ops.py

```diff
@@ -239,21 +239,25 @@ def __init__(self, block: AbstractBlock, n_qubits: int, config: Configuration =
             config = Configuration()
         self.n_qubits = n_qubits
         if block._is_diag_pauli and not block.is_parametric:
-            diag = block_to_diagonal(block, tuple(range(n_qubits)))
-            self.register_buffer("diag", diag)
+            self.register_buffer(
+                "diagonal_observable", block_to_diagonal(block, tuple(range(n_qubits)))
+            )
 
-            def sparse_operation(state: Tensor, values: dict[str, Tensor] = None) -> Tensor:
-                return pyqify(diag * unpyqify(state), n_qubits=self.n_qubits)
+            def _forward(
+                pyq_obs: PyQObservable, state: Tensor, values: dict[str, Tensor] = None
+            ) -> Tensor:
+                return pyqify(pyq_obs.diagonal_observable * unpyqify(state), n_qubits=self.n_qubits)
 
-            self.operation = sparse_operation
+            self._forward = _forward
```

Comment:
hmmm? I'm confused

Reply:
Ok, looking at the whole file I think I understand. It's not very pretty :p

I guess I find it strange to make an actual function rather than a lambda if we first do this.

Reply:
yeah its far from pretty

Reply:
improved it now